integration_test.go

```diff
@@ -0,0 +1,138 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	cli "gopkg.in/urfave/cli.v1"
+)
+
+func TestFetchWithBranchOption(t *testing.T) {
+	t.Parallel()
+
+	tmpDownloadPath := createTempDir(t, "fetch-branch-test")
+
+	cases := []struct {
+		name       string
+		repoUrl    string
+		branchName string
+		sourcePath string
+	}{
+		// Test on a public repo whose sole purpose is to be a test fixture for this tool
+		{"branch option with public repo", "https://github.com/gruntwork-io/fetch-test-public", "sample-branch", "/"},
+
+		// Private repo equivalent
+		{"branch option with private repo", "https://github.com/gruntwork-io/fetch-test-private", "sample-branch", "/"},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			cmd := fmt.Sprintf("fetch --repo %s --branch %s --source-path %s %s", tc.repoUrl, tc.branchName, tc.sourcePath, tmpDownloadPath)
+			//t.Fatal(cmd)
+			output, _, err := runFetchCommandWithOutput(t, cmd)
+			require.NoError(t, err)
+
+			// When --branch is specified, ensure the latest commit is fetched
+			assert.Contains(t, output, "Downloading latest commit from branch")
```

Comment:
Is there some file in that branch we could check instead of `stdout`? Might simplify the test and make it more robust against minor logging changes.

Reply:
@brikis98 Yeah, there are two files specifically added in that branch in both repos, BUT if that branch exists with a particular Git tag, then this test might pass with an older version of the said branch. We don't want to allow that behavior as we want the latest commit. I tried to only assert for a subset of the output to future-proof it. Maybe I could look up the latest commit manually or hard-code it as the branches haven't changed in 5 years?

Reply:
Perhaps create a totally new branch, specifically for testing? Call it something like `branch-for-testing`. There should be no tag of that name or any other conflict to worry about.

Reply:
(Create a totally new branch and totally new file in that branch, that is)

Reply:
@brikis98 we'll we kind of already have that, see: https://github.com/gruntwork-io/fetch-test-public/tree/sample-branch. Two [files get added](https://github.com/gruntwork-io/fetch-test-public/compare/sample-branch). I'm of the opinion we should actually assert for both the text as I've done and whether or not the file exists.

Because we can continue to use the new integration helper methods like `runFetchCommandWithOutput` until we eventually move our `fetch` code over to use an inbuilt logger.

Reply:
Eventually, we can add more integration tests for other use cases.

Reply:
Ah, if we already have those files in such a branch, then I think checking for those files is enough... We don't need the fancy `stdout` check. We could even run the code explicitly against `master` branch and than this `sample-branch` and ensure the files you see are different!

Reply:
@brikis98 yeah, I previously ran it against `master` as it will always exist but I noticed Josh created these branches that look like they are exclusively used for testing. The only problem I have with removing the `stdout` check is whilst it's not ideal, we miss the opportunity to write further integration tests in the future. I'm of the opinion it would be better to leave functions like `runFetchCommand` and `runFetchCommandWithOutput` in, to properly test command output. In the long run, I think the right solution is to refactor the code to avoid writing directly to those streams and then removing the hacks in the test suite.

Reply:
For instance, you can see how we do this properly in Terragrunt: https://github.com/gruntwork-io/terragrunt/blob/master/test/integration_test.go

Reply:
I think the real solution is to switch `fetch` to use a proper logger. That way, we'd have log levels, logs would go to `stderr`, we could swap in a logger more easily that captures the output in tests, and so on. But that's a fair bit of work.

Your call on whether to leave the `stdout` check that's there now or not! I don't feel strongly either way.

Reply:
Yes, I think we're in alignment now. I'm not sure if you missed the fact I mentioned a logger above:

> Because we can continue to use the new integration helper methods like runFetchCommandWithOutput until we eventually move our fetch code over to use an inbuilt logger.

I'm under the opinion that this is non-ideal, but moving in the right direction, so I'm going to leave it as is, but improve the test cases and mention why I'm not running the tests in parallel. Eventually, we can introduce a logger.

Reply:
Resolving this one, but opened a separate ticket to track the issue: https://github.com/gruntwork-io/fetch/issues/89

Reply:
fixed in latest commit.